### PR TITLE
[WIP] Implement read jobType to simulate load with GET requests

### PR DESCRIPTION
**Pull request by riuvshyn**
## Type of change

Attempt to address https://github.com/kube-burner/kube-burner/issues/575

- [ ] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] Optimization
- [x] Documentation Update

## Description

Adds new `jobType: read` which can be used to simulate load of GET requests 

## Related Tickets & Documents

- Related Issue #
- Closes #

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.

### CODE REVIEW

File: pkg/burner/read.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+		}
+		if len(o.LabelSelector) == 0 {
+			log.Fatalf("Empty labelSelectors not allowed with: %s", o.Kind)
+		}
+		obj := object{
+			gvr:           mapping.Resource,
+			labelSelector: o.LabelSelector,
+		}
+		obj.Namespaced = mapping.Scope.Name() == meta.RESTScopeNameNamespace
+		log.Infof("Job %s: %d iterations", jobConfig.Name, jobConfig.JobIterations)
```
Comment (rsevilla87) [inline pkg/burner/read.go:52]: I think is better to take this message outside the loop

File: pkg/burner/read.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+		log.Debugf("Job %s: Read %s with selector %s", jobConfig.Name, gvk.Kind, labels.Set(obj.labelSelector))
+		ex.objects = append(ex.objects, obj)
+	}
+	return ex
+}
+
+// RunReadJob executes a reading job
+func (ex *Executor) RunReadJob(iterationStart, iterationEnd int) {
+	var wg sync.WaitGroup
```
Comment (rsevilla87) [inline pkg/burner/read.go:61]: A workgroup is not required here, the rate limiter already takes care of limiting the number of requests in parallel

**Review by rsevilla87 — CHANGES_REQUESTED** [review]
This feature is quite cool, simple but effective. Now I understand your desire of running jobs in parallel...

However I have some doubts, when the number of objects is high, that will span as many goroutines as the number of objects listed.

I wonder if we can limit the number of concurrent goroutines to QPS to prevent potential issues. (just realized that same issue can happen in delete jobs)

Also, can you a update tests to include this new feature?

File: pkg/burner/read.go (hunk 1 above)
Comment (riuvshyn) [reply pkg/burner/read.go:52]: absolutely, missed that bit.

File: pkg/burner/read.go (hunk 2 above)
Comment (riuvshyn) [reply pkg/burner/read.go]: oh, very nice!
done.

**Comment by riuvshyn** [comment]
@rsevilla87 I was thinking about problem with high amount of objects spawning many goroutines, limiting amount of concurrent goroutines to QPS probably is not very good idea because based on my use cases I have not that many objects defined in my job while in some jobs I am tempting to set quite high QPS to produce significant load which I think is not unique use case for kube-burner.
I am not very concerned with creating goroutines for every object as I can't imagine sane use-case with hundreds of objects being involved, while setting up QPS to hundreds or even thousands makes alot of sense to me.

File: pkg/burner/read.go (hunk 1 above)
Comment (riuvshyn) [reply pkg/burner/read.go]: done

File: pkg/burner/read.go (hunk 2 above)
Comment (rsevilla87) [reply pkg/burner/read.go]: My bad, we need a waitGroup to wait for goroutines to finish before moving forward to the next iteration. sorry about the confusion, I think the initial code was missing the statement wg.Wait()

**Comment by rsevilla87** [comment]
> @rsevilla87 I was thinking about problem with high amount of objects spawning many goroutines, limiting amount of concurrent goroutines to QPS probably is not very good idea because based on my use cases I have not that many objects defined in my job while in some jobs I am tempting to set quite high QPS to produce significant load which I think is not unique use case for kube-burner. I am not very concerned with creating goroutines for every object as I can't imagine sane use-case with hundreds of objects being involved, while setting up QPS to hundreds or even thousands makes alot of sense to me.

ok, makes sense, let's give it a try, we can always update and improve it in the future. Can you add a simple test case to https://github.com/kube-burner/kube-burner/blob/main/test/test-k8s.bats

Thanks!

**Comment by riuvshyn** [comment]
@riuvshyn I've added simple read test case seems it is passing on CI, I was unable to run it locally because I don't have required local infra.

**Comment by rsevilla87** [comment]
> @rsevilla87 I've added simple read test case seems it is passing on CI, I was unable to run it locally because I don't have required local infra.

Great!, thank you very much!

**Review by rsevilla87 — APPROVED** [review]
lgtm!

**Pull request merged**